{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import random\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import os \n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "Chromosome_1_ = 'Final_bed_files/Chromosomes/Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa'\n",
    "Chromosome_2_ = 'Final_bed_files/Chromosomes/Arabidopsis_thaliana.TAIR10.dna.chromosome.2.fa'\n",
    "Chromosome_3_ = 'Final_bed_files/Chromosomes/Arabidopsis_thaliana.TAIR10.dna.chromosome.3.fa'\n",
    "Chromosome_4_ = 'Final_bed_files/Chromosomes/Arabidopsis_thaliana.TAIR10.dna.chromosome.4.fa'\n",
    "Chromosome_5_ = 'Final_bed_files/Chromosomes/Arabidopsis_thaliana.TAIR10.dna.chromosome.5.fa'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence: CCCTAAACCCTAAACCCTAAACCCTAAACCTCTGAATCCTTAATCCCTAA ...\n",
      "Sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN ...\n",
      "Sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN ...\n",
      "Sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN ...\n",
      "Sequence: TATACCATGTACCCTCAACCTTAAAACCCTAAAACCTATACTATAAATCT ...\n"
     ]
    }
   ],
   "source": [
    "# Putting all the chromosomes in a list with a size of 5\n",
    "seqs_List= []\n",
    "\n",
    "file_paths = [Chromosome_1_,Chromosome_2_,Chromosome_3_,Chromosome_4_,Chromosome_5_]\n",
    "\n",
    "# Iterate over each file\n",
    "for file_path in file_paths:\n",
    "   with open(file_path, \"r\") as handle:\n",
    "\n",
    "    # Open the file\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "            sequence_id = record.id\n",
    "            #print(\"Sequence ID:\", sequence_id)\n",
    "            description = record.description\n",
    "            #print(\"Description:\", description)\n",
    "            sequence = record.seq\n",
    "            seqs_List.append(str(sequence))\n",
    "            print(\"Sequence:\", str(sequence[:50]), \"...\")  # Print first 50 characters of the sequence\n",
    "            #print(\"Sequence length:\", len(sequence))  # Print length of the sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_dataframe(filename):\n",
    "    # Read the CSV file into a DataFrame\n",
    "    df = pd.read_csv(filename, sep='\\t', header=None)\n",
    "\n",
    "    # Set column names\n",
    "    df.columns = ['chromosome', 'start', 'end', 'name', 'score', 'strand', 'stuff', 'things', 'who', 'ahh', 'scream', 'ska', 'die', 'grr']\n",
    "\n",
    "    # Calculate new positions based on existing start and end positions\n",
    "    df['dif'] = df['end'] - df['start']\n",
    "    df['mid'] = df['dif'] // 2\n",
    "    df['midpoint'] = df['start'] + df['mid']\n",
    "    df['new_start'] = df['midpoint'] - 400\n",
    "    df['new_end'] = df['midpoint'] + 400\n",
    "    df['dif_new'] = df['new_end'] - df['new_start']\n",
    "\n",
    "    # Create a new DataFrame with selected columns\n",
    "    df_new = pd.DataFrame({\n",
    "        'chromosome': df['chromosome'],\n",
    "        'start': df['start'],\n",
    "        'end': df['end'],\n",
    "        'midpoint': df['midpoint'],\n",
    "        'dif': df['dif'],\n",
    "        'new_start': df['new_start'], #new start and end are from the midpoint out 250 each way\n",
    "        'new_end': df['new_end'],\n",
    "        'position_difference_new': df['dif_new']\n",
    "    })\n",
    "\n",
    "    return df_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "testFile = \"Final_bed_files/SRP034156_BedFiles/SRX391990.target.all.bed\"\n",
    "directory = \"Final_bed_files//test_bedFiles\"\n",
    "\n",
    "# List to store dataframes\n",
    "dataframes = []\n",
    "\n",
    "for filename in os.listdir(directory):\n",
    "    if filename.endswith(\".bed\"):\n",
    "        file_path = os.path.join(directory, filename)\n",
    "        df = process_dataframe(file_path)\n",
    "        dataframes.append(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_seqs(df, data_list):\n",
    "    dfs = []\n",
    "    for index, row in df.iterrows():\n",
    "        #Changing the chromosome to an int that can be used to find the correct index of the string of chromosomes\n",
    "        chromosome_label = row['chromosome']\n",
    "        chromosome_number = int(chromosome_label.replace('Chr', ''))\n",
    "        data = data_list[chromosome_number - 1]\n",
    "        extracted_data = data[row['new_start']:row['new_end']] #gets the peak sequence\n",
    "        size = len(extracted_data)\n",
    "        \n",
    "        if(size < 800):\n",
    "            Add_num_N = 800 - size\n",
    "            #print(size)\n",
    "            print(\"N to add: \", Add_num_N)\n",
    "            extracted_data += 'N' * Add_num_N\n",
    "            new_size = len(extracted_data)\n",
    "            #print(new_size)\n",
    "            size = new_size\n",
    "\n",
    "        df_extracted = pd.DataFrame({ #saves it as a new dataframe\n",
    "            'chromosome': [chromosome_label],\n",
    "            'start': [row['new_start']],\n",
    "            'end': [row['new_end']],\n",
    "            'size': [size],\n",
    "            'data': [extracted_data]\n",
    "        })\n",
    "\n",
    "        dfs.append(df_extracted)\n",
    "\n",
    "    result_df = pd.concat(dfs, ignore_index=True)\n",
    "    print(\"done\")\n",
    "    return result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame 1: 29238 rows x 8 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'midpoint', 'dif', 'new_start', 'new_end', 'position_difference_new']\n",
      "DataFrame 2: 13905 rows x 8 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'midpoint', 'dif', 'new_start', 'new_end', 'position_difference_new']\n",
      "DataFrame 3: 16752 rows x 8 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'midpoint', 'dif', 'new_start', 'new_end', 'position_difference_new']\n",
      "DataFrame 4: 19464 rows x 8 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'midpoint', 'dif', 'new_start', 'new_end', 'position_difference_new']\n",
      "DataFrame 5: 20457 rows x 8 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'midpoint', 'dif', 'new_start', 'new_end', 'position_difference_new']\n",
      "DataFrame 6: 24019 rows x 8 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'midpoint', 'dif', 'new_start', 'new_end', 'position_difference_new']\n",
      "DataFrame 7: 22932 rows x 8 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'midpoint', 'dif', 'new_start', 'new_end', 'position_difference_new']\n",
      "DataFrame 8: 11987 rows x 8 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'midpoint', 'dif', 'new_start', 'new_end', 'position_difference_new']\n",
      "DataFrame 9: 22100 rows x 8 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'midpoint', 'dif', 'new_start', 'new_end', 'position_difference_new']\n",
      "DataFrame 10: 13879 rows x 8 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'midpoint', 'dif', 'new_start', 'new_end', 'position_difference_new']\n"
     ]
    }
   ],
   "source": [
    "# Check size and column names of each DataFrame within the list\n",
    "for i, df in enumerate(dataframes):\n",
    "    num_rows, num_columns = df.shape\n",
    "    print(f\"DataFrame {i+1}: {num_rows} rows x {num_columns} columns\")\n",
    "    print(f\"Columns: {df.columns.tolist()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N to add:  246\n",
      "done\n",
      "Columns in processed DataFrame: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "done\n",
      "Columns in processed DataFrame: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "done\n",
      "Columns in processed DataFrame: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "N to add:  37\n",
      "done\n",
      "Columns in processed DataFrame: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "done\n",
      "Columns in processed DataFrame: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "done\n",
      "Columns in processed DataFrame: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "N to add:  44\n",
      "done\n",
      "Columns in processed DataFrame: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "done\n",
      "Columns in processed DataFrame: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "done\n",
      "Columns in processed DataFrame: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "done\n",
      "Columns in processed DataFrame: ['chromosome', 'start', 'end', 'size', 'data']\n"
     ]
    }
   ],
   "source": [
    "Sequences_DataFrame = []\n",
    "\n",
    "for k in dataframes:\n",
    "    processed_df = add_seqs(k, seqs_List)\n",
    "    Sequences_DataFrame.append(processed_df)\n",
    "    # Print the columns of the DataFrame after processing\n",
    "    print(\"Columns in processed DataFrame:\", processed_df.columns.tolist())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame 1: 29238 rows x 5 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "DataFrame 2: 13905 rows x 5 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "DataFrame 3: 16752 rows x 5 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "DataFrame 4: 19464 rows x 5 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "DataFrame 5: 20457 rows x 5 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "DataFrame 6: 24019 rows x 5 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "DataFrame 7: 22932 rows x 5 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "DataFrame 8: 11987 rows x 5 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "DataFrame 9: 22100 rows x 5 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'size', 'data']\n",
      "DataFrame 10: 13879 rows x 5 columns\n",
      "Columns: ['chromosome', 'start', 'end', 'size', 'data']\n"
     ]
    }
   ],
   "source": [
    "# Check size and column names of each DataFrame within the list\n",
    "for i, df in enumerate(Sequences_DataFrame):\n",
    "    num_rows, num_columns = df.shape\n",
    "    print(f\"DataFrame {i+1}: {num_rows} rows x {num_columns} columns\")\n",
    "    print(f\"Columns: {df.columns.tolist()}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function generates labeled datasets by appending both original (positive) and shuffled (negative) DNA sequences, ensuring all sequences are of a specified length, and includes chromosome information.\n",
    "def generate_labels_and_split_all(dataframes_list, data_list, seq_length=800):\n",
    "    all_labeled_data = []\n",
    "    for df in dataframes_list:\n",
    "        labeled_data = []\n",
    "        for index, row in df.iterrows():\n",
    "            chromosome_label = row['chromosome']  # Storing the original chromosome label\n",
    "            chromosome_number = int(chromosome_label.replace('Chr', ''))\n",
    "            # Use 'start' and 'end' directly\n",
    "            sequence = data_list[chromosome_number - 1][row['start']:row['end']]\n",
    "            \n",
    "            # Ensure the sequence is of required length\n",
    "            if len(sequence) == seq_length:\n",
    "                labeled_data.append({'chromosome': chromosome_label, 'sequence': sequence, 'label': 1})  # Positive example\n",
    "                # Generating a negative example by shuffling\n",
    "                negative_sequence = ''.join(random.sample(sequence, len(sequence)))\n",
    "                labeled_data.append({'chromosome': chromosome_label, 'sequence': negative_sequence, 'label': 0})  # Negative example\n",
    "        \n",
    "        if labeled_data:\n",
    "            all_labeled_data.append(pd.DataFrame(labeled_data))\n",
    "    \n",
    "    return pd.concat(all_labeled_data, ignore_index=True)\n",
    "\n",
    "# Generate the labeled DataFrame\n",
    "labeled_df = generate_labels_and_split_all(Sequences_DataFrame, seqs_List)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Labeled DataFrame: 389460 rows x 3 columns\n",
      "Columns: ['chromosome', 'sequence', 'label']\n",
      "chromosome                                                 Chr1\n",
      "sequence      TAAGAAATCCATACATCAACATATCGCTTTCGTTACCTTAAATTTT...\n",
      "label                                                         1\n",
      "Name: 0, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# Check size and column names of the labeled DataFrame\n",
    "num_rows, num_columns = labeled_df.shape\n",
    "print(f\"Labeled DataFrame: {num_rows} rows x {num_columns} columns\")\n",
    "print(f\"Columns: {labeled_df.columns.tolist()}\")\n",
    "print(labeled_df.iloc[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame encoded. First encoded sequence shape: (800, 5)\n"
     ]
    }
   ],
   "source": [
    "# Function to convert a sequence to one-hot encoding\n",
    "def one_hot_encode(seq):\n",
    "    # Mapping of nucleotides to integers\n",
    "    mapping = {'A': 0, 'C': 1, 'G': 2, 'T': 3, 'N': 4}\n",
    "    # Create a zero matrix of size (len(seq), 5)\n",
    "    one_hot = np.zeros((len(seq), 5), dtype=np.int8)\n",
    "    # Fill the appropriate positions\n",
    "    for i, nucleotide in enumerate(seq):\n",
    "        if nucleotide in mapping:\n",
    "            one_hot[i, mapping[nucleotide]] = 1\n",
    "    return one_hot\n",
    "\n",
    "# Applying one-hot encoding to extracted DNA sequences\n",
    "def encode_sequences(df):\n",
    "    # Applying the one-hot encoding to each row in the dataframe using the 'sequence' column\n",
    "    df['one_hot_data'] = df['sequence'].apply(one_hot_encode)\n",
    "    return df\n",
    "\n",
    "# Apply encoding to the labeled DataFrame\n",
    "encoded_df = encode_sequences(labeled_df)\n",
    "print(\"DataFrame encoded. First encoded sequence shape:\", encoded_df['one_hot_data'].iloc[0].shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome: Chr1\n",
      "Label: 0\n",
      "DNA Sequence: GAAGTACTAAACCACAATCTCTGTATCTGTACGTCAATACAACGACTCATCGACATTCGGTCAAAATGGATATGTGTAACTGGATGAAAATCGCAACTAAAGTGAAAAGAACTTACGCATGCCTAAGTATTCCTAAAGTGATAGTCACATTTAAAATGAATCATGCTGTCCACCGTCGTATAATGAATGAGTGTATATAATGCAGTTAATGAATATCACTGTATATGTCACTATAAGCGACCGTGGCAAACACAGTTTATTTTAAACGACGGTTTACGATGCGAACACCAATCACACAGACTTGTCCATAGAAAATCTTAGCGCCAATCCCGTAAAGAATTATTCTCTGAAGAGATTGAAAAGCTACCACCTAGTTTATACAATCCTGTAAAGCTAGTAATTAATGTCAGGTAAAATTGTTATATGAATCTAATCAGTTTGGCGGAGTAACTCAGCGATAAGCACCGACATTTAACACTCTAGGGTAGTCGTGCTAGCTTTTGATTATAAATATATAAAGCTATTTATACGGAGTTATATATTTATAAACCAGAATCAGAGAAGACGTGGATTTATCATCACTTAAGCTATGACATCAGATGCTATAGTTTTCGATAAACCACTGTTTGACAACATAACCGACATGCCATTTGCCAGGCATATATCTTATTCCACAATATCTACACCAAAAAATCCTAACGTTAGATCGTATGTACGTTTCAGTTACACTAGATTCAAATGGATGAGTTCTGCCTACTTCAATACCTATGACCTTAACATGTCAAATTAACCCGATTTTT\n",
      "One-hot Encoded Sequence:\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 1 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 1 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[1 0 0 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 1 0 0 0]\n",
      "[0 0 1 0 0]\n",
      "[1 0 0 0 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n",
      "[0 0 0 1 0]\n"
     ]
    }
   ],
   "source": [
    "# Function to print an entry from the dataframe, just used to verify one hot function worked\n",
    "def print_one_hot_entry(df, index):\n",
    "    if index < len(df):\n",
    "        # Extracting data for the specified index\n",
    "        entry = df.iloc[index]\n",
    "        chromosome = entry['chromosome']\n",
    "        sequence = entry['sequence']\n",
    "        label = entry['label']  # Retrieving the label\n",
    "        one_hot_sequence = entry['one_hot_data']\n",
    "\n",
    "        # Printing the details\n",
    "        print(f\"Chromosome: {chromosome}\")\n",
    "        print(f\"Label: {label}\")\n",
    "        print(f\"DNA Sequence: {sequence}\")\n",
    "        print(\"One-hot Encoded Sequence:\")\n",
    "        for row in one_hot_sequence:  # Printing each row of the one-hot encoded sequence\n",
    "            print(row)\n",
    "    else:\n",
    "        print(\"Index out of range.\")\n",
    "\n",
    "# Assuming encoded_df is the DataFrame you want to check\n",
    "print_one_hot_entry(encoded_df, 1)  # Print the first entry of the labeled and encoded DataFrame\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame saved to encoded_df.pkl\n"
     ]
    }
   ],
   "source": [
    "# Function to save DataFrame to a pickle file\n",
    "def save_encoded_df(df, filename):\n",
    "    with open(filename, 'wb') as f:\n",
    "        pickle.dump(df, f)\n",
    "    print(f\"DataFrame saved to {filename}\")\n",
    "\n",
    "# Save the encoded DataFrame\n",
    "save_encoded_df(encoded_df, 'encoded_df.pkl')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
